### PR TITLE
io: functions for multiplication tables

### PR DESCRIPTION
**Pull request by ChristopherRussell**
Add three new functions for reading and writing multiplication tables from files:

* `ReadMultiplicationTable`
* `WriteMultiplicationTable`
* `IteratorFromMultiplicationTable`

These functions can be used to access a database of all the inverse semigroups of order less than or equal to 12 (this may rise to 13 or 14). The relevant files will soon be available [here](http://www-groups.mcs.st-andrews.ac.uk/~jamesm/data.php)

### CODE REVIEW

File: data/tst/tables [hunk 1]
```diff
@@ -0,0 +1,48 @@
+0023456789012345678922345678903345678902445678902355678902346678902345778902345688902345679902345678
```
Comment (james-d-mitchell) [inline data/tst/tables]: You should gzip this file.

File: doc/io.xml [hunk 2]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
     </Description>
   </ManSection>
 <#/GAPDoc>
+
+<#GAPDoc Label="IteratorFromMultiplicationTable">
```
Comment (james-d-mitchell) [inline doc/io.xml]: I think this function requires the suffix "File" otherwise the name is misleading.

File: doc/io.xml [hunk 3]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+      If <A>filename</A> is a string containing the name of a file created using
+      <Ref Func = "WriteMultiplicationTable"/>, then 
+      <C>IteratorFromMultiplicationTable</C> returns an iterator <C>iter</C>
+      such that <C>NextIterator(iter)</C> returns the next multiplication table
+      stored in the file <A>filename</A>.<P/>
+
+      This function is a convenient way of, for example, looping over a
+      collection of multiplication tables in a file without loading every
+      object in the file into memory. This might be useful if the file
+      contains more information than there is available memory.<P/>
```
Comment (james-d-mitchell) [inline doc/io.xml]: No new paragraph

File: doc/io.xml [hunk 4]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+      The argument <A>filename</A> should be a string containing the name of a
+      file where the entries in <A>list</A> will be written or an
+      <Package>IO</Package> package file object; see <Ref Func="IO_File"
+        BookName="io"/> and <Ref Func="IO_CompressedFile" BookName="io"/>.<P/>
+
+      If  the  optional  third argument <A>append</A> is given and equals
+      <C>"w"</C> or is not present, then the previous content of the file
+      is deleted and overwritten. If the optional third argument is <C>"a"</C>
+      or is not present, then <C>list</C> is appended to the file. This
+      function returns <Ref Var="IO_OK" BookName="io"/> if everything went well
```
Comment (james-d-mitchell) [inline doc/io.xml]: So, if the optional third argument is not present, what happens?

File: doc/io.xml [hunk 5]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+      or <Ref Var="IO_ERROR" BookName="io"/> if something went wrong.<P/>
+
+      <C>WriteMultiplicationTable</C> appends a line to the file
```
Comment (james-d-mitchell) [inline doc/io.xml]: This contradicts the previous paragraph, where it says that sometimes it appends and other times overwrites.

File: doc/io.xml [hunk 6]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+      <A>filename</A> for every entry in <A>list</A>.<P/>
+      
+      The file <A>filename</A> can be read using
```
Comment (james-d-mitchell) [inline doc/io.xml]: Maybe "The multiplication tables saved in <A>filename</A> can be recovered from the file using ..."

File: doc/io.xml [hunk 7]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+
+<#GAPDoc Label="ReadMultiplicationTable">
+  <ManSection>
+    <Func Name = "ReadMultiplicationTable" Arg = "filename[, nr]"/>
+    <Returns>A list of multiplication tables.</Returns>
+    <Description>
+      If <A>filename</A> is the name of a file created using
+      <Ref Func = "WriteMultiplicationTable"/>, then
+      <C>ReadMultiplicationTable</C> returns the contents of this file as a
+      list of multiplication tables<P/>
```
Comment (james-d-mitchell) [inline doc/io.xml]: Missing full stop.

File: doc/io.xml [hunk 8]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+
+      If the optional second argument <A>nr</A> is present, then
+      <C>ReadMultiplicationTable</C> returns the mutiplication table stored in
+      the <A>nr</A>th line of <A>filename</A>.<P/>
```
Comment (james-d-mitchell) [inline doc/io.xml]: No new paragraph.

File: doc/io.xml [hunk 9]
```diff
@@ -105,3 +105,80 @@ gap> ReadGenerators(file, 13);
...
+      
+      The file <A>filename</A> can be read using 
+      <Ref Func = "ReadMultiplicationTable"/>.<P/>
+    </Description>
+  </ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="ReadMultiplicationTable">
+  <ManSection>
+    <Func Name = "ReadMultiplicationTable" Arg = "filename[, nr]"/>
```
Comment (james-d-mitchell) [inline doc/io.xml]: This also doesn't mention that `filename` can be a file object, if this is the case, then this should be added to the description.

File: gap/tools/io.gi [hunk 10]
```diff
@@ -126,7 +126,7 @@ function(arg)
   if Length(arg) = 2 then
     name := arg[1];
     coll := arg[2];
-    mode := "w";
+    mode := "w"; #FIXME: Doc says this should be "a" by default.
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: If the file doesn't exist then having mode "a" fails. Maybe this is related?

File: gap/tools/io.gi [hunk 11]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+  local name, line_nr, file, ReadMultiplicationTableLine, i, line, n, lines;
+  if Length(arg) = 1 then
+    name    := arg[1];
+    line_nr := 0;
+  elif Length(arg) = 2 then
+    name    := arg[1];
+    line_nr := arg[2];
+  else
+    ErrorNoReturn("Semigroups: ReadMultiplicationTable: usage,\n",
+                  "there should be at most 2 arguments,");
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: "there should be 1 or 2 arguments" since this error will be shown if there are no arguments.

File: gap/tools/io.gi [hunk 12]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+      i    := i + 1;
+      line := IO_ReadLine(file);
+    until i = line_nr or line = "";
+    if IsString(arg[1]) then
+      IO_Close(file);
+    elif line = "" then
+      ErrorNoReturn("Semigroups: ReadMultiplicationTable:\n",
+                    "the file only has ", i - 1, " lines,");
+    fi;
+    n := RootInt(Length(line)); #TODO: Returns the int part. Change or check?
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: Maybe just add some checks here that length of the line is indeed a perfect square

File: gap/tools/io.gi [hunk 13]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+  if not (mode = "a" or mode = "w") then
+    ErrorNoReturn("Semigroups: WriteMultiplcationTable: usage,\n",
+                  "the third argument must be \"a\" or \"w\",");
+  fi;
+
+  if IsString(name) then
+    name := UserHomeExpand(name);
+    file := IO_CompressedFile(name, mode);
+    if file = fail then
+      # Cannot test this
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: Sure you can, just put in a name of a file that doesn't exist

File: gap/tools/io.gi [hunk 14]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+    ErrorNoReturn("Semigroups: WriteMultiplicationTable: usage,\n",
+                  "the first argument must be a string or a file,");
+  fi;
+
+  for i in [1 .. Length(coll)] do
+    n := Size(coll[i]);
+    if not n < 17 then
+      ErrorNoReturn("Semigroups: WriteMultiplicationTable: usage,\n",
+                    "the second argument must be a multiplication table ",
+                    "for a semigroup of order less than 17,");
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: You also do not mention this limitation in the documentation.

File: gap/tools/io.gi [hunk 15]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+              if n > 15 then
+                str := ReplacedString(str, "15", "F");
+              fi;
+            fi;
+          fi;
+        fi;
+      fi;
+    fi;
+    RemoveCharacters(str, " ,[]");
+    if IO_Write(file, str) = fail then #TODO: IO_Pickle?
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: No, `IO_Pickle`, remove this comment.

File: gap/tools/io.gi [hunk 16]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+  for i in [1 .. Length(coll)] do
+    n := Size(coll[i]);
+    if not n < 17 then
+      ErrorNoReturn("Semigroups: WriteMultiplicationTable: usage,\n",
+                    "the second argument must be a multiplication table ",
+                    "for a semigroup of order less than 17,");
+    fi;
+
+    str := String(coll[i] - 1);
+    if n > 10 then
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: This does not look very good, the complexity of this approach (creating a string, and then removing bits and modifying it) is too high, and there is too much code. Why not do something like: 
```str := "";
for j in [1 .. Length(coll[i])] do 
  for k in [1 .. Length(coll[i])] do 
    Append(str, HexStringInt(coll[i][j][k] - 1));
  od;
od;
```

File: gap/tools/io.gi [hunk 17]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+  local file, line, record, n;
+
+  file := IO_CompressedFile(UserHomeExpand(str), "r");
+
+  if file = fail then
+    return fail;
+  fi;
+
+  line := IO_ReadLine(file);
+  n := RootInt(Length(line)); #TODO: Change or check?
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: check

File: gap/tools/io.gi [hunk 18]
```diff
@@ -313,3 +313,212 @@ function(str)
...
+    local file, line;
+    file := IO_CompressedFile(UserHomeExpand(str), "r");
+    line := SEMIGROUPS.ReadMultiplicationTableLine(Chomp(IO_ReadLine(file)), n);
+    return rec(file := file, current := line);
+  end;
+
+  return IteratorByFunctions(record);
+end);
+
+# TODO 1. WriteMultiplicationTable (see WriteGenerators)
```
Comment (james-d-mitchell) [inline gap/tools/io.gi]: Remove this list too

File: gap/tools/io.gi (hunk 10 above)
Comment (ChristopherRussell) [inline gap/tools/io.gi]: Should I change the doc for `WriteGenerators` to indicate that the default option is "w" ?

File: gap/tools/io.gi (hunk 13 above)
Comment (ChristopherRussell) [inline gap/tools/io.gi]: There is an identical comment in the method for WriteGenerators, which isn't covered by tests. What is the difference?

Comment (james-d-mitchell) [inline gap/tools/io.gi]: You're right, this can't be tested, sorry.

File: gap/tools/io.gi (hunk 10 above)
Comment (james-d-mitchell) [inline gap/tools/io.gi]: Yes, that's probably the best solution at present.

**Comment by ChristopherRussell** [comment]
I have now added the checks (and the corresponding new tests) for the `WriteMultiplicationTable` method, which I forgot to do along with the other changes.

File: doc/io.xml [hunk 19]
```diff
@@ -44,17 +44,17 @@
...
 
-      If  the  optional  third argument <A>append</A> is given and equals
-      <C>"w"</C>,  then the previous content of the file is deleted and
-      overwritten.  If the optional third argument is <C>"a"</C> or is not
-      present, then <C>list</C> is appended to the file.  This function returns
-      <Ref Var="IO_OK" BookName="io"/> if everything went well or <Ref
-        Var="IO_ERROR" BookName="io"/> if something went wrong.<P/>
+      If the optional third argument <A>append</A> is not present or is given
+      and equals <C>"w"</C>, then the previous content of the file is deleted
+      and overwritten.  If the third argument is <C>"a"</C> then <C>list</C> is
```
Comment (james-d-mitchell) [inline doc/io.xml]: Comma before "then"

File: doc/io.xml [hunk 20]
```diff
@@ -105,3 +105,93 @@ gap> ReadGenerators(file, 13);
...
+      If <A>filename</A> is a file or a string containing the name of a file
+      created using <Ref Func = "WriteMultiplicationTable"/>, then 
+      <C>IteratorFromMultiplicationTableFile</C> returns an iterator <C>iter</C>
+      such that <C>NextIterator(iter)</C> returns the next multiplication table
+      stored in the file <A>filename</A>.<P/>
+
+      This function is a convenient way of, for example, looping over a
+      collection of multiplication tables in a file without loading every
+      object in the file into memory. This might be useful if the file
+      contains more information than there is available memory.
```
Comment (james-d-mitchell) [inline doc/io.xml]: Missing <P/> at the end of the line

File: doc/io.xml [hunk 21]
```diff
@@ -105,3 +105,93 @@ gap> ReadGenerators(file, 13);
...
+      list of multiplication tables.<P/>
+
+      If the optional second argument <A>nr</A> is present, then
+      <C>ReadMultiplicationTable</C> returns the mutiplication table stored in
+      the <A>nr</A>th line of <A>filename</A>.
+
+      This function may be used to access files containg the multiplication
+      tables of all the inverse semigroups of order less than or equal to 12.
+      These files may be found at: 
+      http://www-groups.mcs.st-andrews.ac.uk/~jamesm/data.php
```
Comment (james-d-mitchell) [inline doc/io.xml]: Delete the last paragraph completely

File: doc/z-chap19.xml [hunk 22]
```diff
@@ -13,4 +13,20 @@
     <#Include Label = "IteratorFromPickledFile">
   </Section>
 
+  <Section>
+    <Heading>Reading and writing multiplication tables to a file</Heading>
+    The functions <Ref Func = "ReadMultiplicationTable"/> and 
+    <Ref Func = "WriteMultiplicationTable"/> can be used to read or write,
+    respectively, multiplication tables to a file.
```
Comment (james-d-mitchell) [inline doc/z-chap19.xml]: Missing <P/>

File: doc/z-chap19.xml [hunk 23]
```diff
@@ -13,4 +13,20 @@
...
+
+    There are files containg the multiplication tables of all the inverse
+    semigroups of order less than or equal to 12. These files may be found at
+    http://www-groups.mcs.st-andrews.ac.uk/~jamesm/data.php
```
Comment (james-d-mitchell) [inline doc/z-chap19.xml]: Also delete this paragraph

File: doc/z-chap19.xml (hunk 22 above)
Comment (ChristopherRussell) [inline doc/z-chap19.xml]: What is missing?

Comment (james-d-mitchell) [inline doc/z-chap19.xml]: A \<P/\>

Comment (ChristopherRussell) [inline doc/z-chap19.xml]: I am deleting the next paragraph, as advised, so this is now the last one. Can I omit the <P/>?

Comment (james-d-mitchell) [inline doc/z-chap19.xml]: Yup

**Pull request merged**